src/libAtomVM/otp_socket.c

```diff
@@ -0,0 +1,1364 @@
+
+        int sent_data = -1;
+
+        if (is_sendto) {
+
+            struct sockaddr_in destaddr;
+            memset(&destaddr, 0, sizeof(destaddr));
+            destaddr.sin_family = AF_INET;
+
+            term port = interop_kv_get_value_default(dest, port_atom, term_from_int(0), global);
+            destaddr.sin_port = htons(term_to_int(port));
+            term addr = interop_kv_get_value(dest, addr_atom, ctx->global);
+            if (globalcontext_is_term_equal_to_atom_string(global, addr, loopback_atom)) {
+                destaddr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+            } else {
+                destaddr.sin_addr.s_addr = htonl(socket_tuple_to_addr(addr));
+            }
+
+            sent_data = sendto(fd_obj->fd, buf, len, 0, (struct sockaddr *) &destaddr, sizeof(destaddr));
+
+        } else {
+            sent_data = send(fd_obj->fd, buf, len, 0);
+        }
+
+        // {ok, RestData} | {error, Reason}  (RestData currently nil)
+        // TODO we should allocate space for the sub-binary here when we support returning Rest
+        if (UNLIKELY(memory_ensure_free(ctx, TUPLE_SIZE(2)) != MEMORY_GC_OK)) {
+            AVM_LOGW(TAG, "Failed to allocate memory: %s:%i.\n", __FILE__, __LINE__);
+            RAISE_ERROR(OUT_OF_MEMORY_ATOM);
+        }
+
+        term ret = term_invalid_term();
+        if (sent_data == -1) {
+
+            AVM_LOGE(TAG, "Unable to send data: res=%i.", sent_data);
+            term error_tuple = term_alloc_tuple(2, &ctx->heap);
+            term_put_tuple_element(error_tuple, 0, ERROR_ATOM);
+            term_put_tuple_element(error_tuple, 1, globalcontext_make_atom(global, closed_atom));
+            ret = error_tuple;
+
+        } else {
+            // TODO allocate substring that references the data that was not sent
+            ret = port_create_tuple2(ctx, OK_ATOM, term_nil());
+        }
+
+        if (term_is_list(data)) {
+            free(buf);
+        }
+        return ret;
+
+    } else {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+}
+
+static term nif_socket_send(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_socket_send\n");
+    return nif_socket_send_internal(ctx, argc, argv, false);
+}
+
+static term nif_socket_sendto(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_socket_sendto\n");
+    return nif_socket_send_internal(ctx, argc, argv, true);
+}
+
+//
+// connect
+//
+
+static term nif_socket_connect(Context *ctx, int argc, term argv[])
+{
+    TRACE("nif_socket_connect\n");
+    UNUSED(argc);
+
+    VALIDATE_VALUE(argv[0], term_is_socket);
+
+    GlobalContext *global = ctx->global;
+
+    void *fd_obj_ptr;
+    if (UNLIKELY(!enif_get_resource(erl_nif_env_from_context(ctx), get_socket(argv[0]), socket_resource_type, &fd_obj_ptr))) {
+        RAISE_ERROR(BADARG_ATOM);
+    }
+    struct SocketFd *fd_obj = (struct SocketFd *) fd_obj_ptr;
+    if (fd_obj->fd) {
+
+        term sockaddr = argv[1];
+
+        // TODO make connect non-blocking
```

Comment:
❤️

Reply:
Do you think this needs to be addressed in this PR?